### PR TITLE
Update perform valve isolation trace sample

### PR DESCRIPTION
**Pull request by Rachael-E**
This PR updates the Perform Valve Isolation Trace sample to add the option of a user manually adding filter barriers to the utility network, as well as using the existing functionality of choosing the filter barriers from a drop down menu. This PR contains:

- new UI for adding filter barriers to the map (handling a map view click to add the location of a filter barrier as a red cross, and adding that location as a utility element to the utility trace parameter's list of filter barriers.
- new UI for resetting the UI/trace so that a user can hit reset to clear the map view of any added filter barriers/traces
- updated readme for how to use the sample and how it works
- new class for handling UI if clicking on a junction returns more than one terminal (same as the class/functionality used in Trace a Utility Network)
- updated metadata json file
- updated screenshot of the sample

@jenmerritt could you take a look please? thanks!

### CODE REVIEW

File: utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java [hunk 1]
```diff
@@ -16,17 +16,49 @@
...
+import com.esri.arcgisruntime.utilitynetworks.UtilityNetwork;
+import com.esri.arcgisruntime.utilitynetworks.UtilityNetworkDefinition;
+import com.esri.arcgisruntime.utilitynetworks.UtilityNetworkSource;
+import com.esri.arcgisruntime.utilitynetworks.UtilityTerminal;
+import com.esri.arcgisruntime.utilitynetworks.UtilityTier;
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceConfiguration;
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceFilter;
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceParameters;
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceResult;
+import com.esri.arcgisruntime.utilitynetworks.UtilityTraceType;
```
Comment (jenmerritt) [inline utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java]: There are 2 sections of `com.esri.arcgisruntime` imports you could tidy up

File: utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java [hunk 2]
```diff
@@ -303,6 +328,144 @@ private void handleTraceClick() {
     }
   }
 
+  /**
+   *  Uses the clicked map point to identify any utility elements in the utility network at the clicked location. The
+   *  clicked utility element is added either to the utility trace parameter's filter barrier list. A graphic is created
```
Comment (jenmerritt) [inline utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java]:
```suggestion
   *  clicked utility element is added to the utility trace parameter's filter barrier list. A graphic is created
```
Possible correction, or there may be an "either" to add here?

File: utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java [hunk 3]
```diff
@@ -303,6 +328,144 @@ private void handleTraceClick() {
...
+          // get the result of the query
+          List<IdentifyLayerResult> identifyLayerResults = identifyLayerResultsFuture.get();
+
+          // return if no features are identified
+          if (!identifyLayerResults.isEmpty()) {
+            // retrieve the first result and get its contents
+            ArcGISFeature identifiedFeature = (ArcGISFeature) identifyLayerResults.get(0).getElements().get(0);
+
+            // create element from the identified feature
+            var utilityElement = utilityNetwork.createElement(identifiedFeature);
```
Comment (jenmerritt) [inline utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java]: Maybe show this one as a `UtilityElement` if that's the case?

File: utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java [hunk 4]
```diff
@@ -303,6 +328,144 @@ private void handleTraceClick() {
...
+
+            // check if the network source is a junction or an edge
+            if (utilityElement.getNetworkSource().getSourceType() == UtilityNetworkSource.Type.JUNCTION) {
+
+              // check if the feature has a terminal configuration and multiple terminals
+              if (utilityElement.getAssetType().getTerminalConfiguration() != null) {
+                var utilityTerminalConfiguration = utilityElement.getAssetType().getTerminalConfiguration();
```
Comment (jenmerritt) [inline utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java]: Again maybe better to do `UtilityTerminalConfiguration` just to make it explicit for the sample

File: utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java [hunk 5]
```diff
@@ -303,6 +328,144 @@ private void handleTraceClick() {
...
+            filterBarriersGraphicsOverlay.getGraphics().add(traceLocationGraphic);
+          }
+
+        } catch (InterruptedException | ExecutionException ex) {
+          statusLabel.setText("Error identifying clicked features.");
+          new Alert(Alert.AlertType.ERROR, "Error identifying clicked features.").show();
+        } finally {
+          progressIndicator.setVisible(false);
+        }
+
```
Comment (jenmerritt) [inline utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java]:
```suggestion
```

File: utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java [hunk 6]
```diff
@@ -312,10 +475,35 @@ private void handleTraceClick() {
...
+    resetButton.setDisable(!enable);
     includeIsolatedFeaturesCheckbox.setDisable(!enable);
   }
 
+  /**
+   * Clears the selection of feature layers from the map, clears the utility trace parameter's list of filter barriers
+   * and clears the graphics within the filter barriers graphics overlay.
+   */
+  @FXML
+  private void handleResetButtonClick() {
```
Comment (jenmerritt) [inline utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java:470]: A final tidy up here could be to reset the text or at least clear out "Isolation trace completed"

**Review by jenmerritt — APPROVED** [review]
The changes look good to me @Rachael-E . Few small suggestions.

**Comment by Rachael-E** [comment]
thanks @jenmerritt ! I've fixed those things up and have also added in additional logic: previously the helper dialogue didn't show terminal options on a node so I've had to dig a little deeper. not sure if this is the best way of doing it, if you had any other suggestions please let me know!

File: utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java [hunk 7]
```diff
@@ -363,68 +341,74 @@ private void handleMapViewClicked(MouseEvent e) {
...
                 }
-              }
 
-            } else if (utilityElement.getNetworkSource().getSourceType() == UtilityNetworkSource.Type.EDGE) {
+              } else if (utilityElement.getNetworkSource().getSourceType() == UtilityNetworkSource.Type.EDGE) {
 
-              // get the geometry of the identified feature as a polyline, and remove the z component
-              Polyline polyline = (Polyline) GeometryEngine.removeZ(identifiedFeature.getGeometry());
+                // get the geometry of the identified feature as a polyline, and remove the z component
+                Polyline polyline = (Polyline) GeometryEngine.removeZ(listOfFeatures.get(0).getGeometry());
```
Comment (jenmerritt) [inline utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java:386]: Is it always going to be the first one? `listOfFeatures.get(0)`

**Review by jenmerritt — COMMENTED** [review]
Looks ok to me @Rachael-E - just one thing to double check

File: utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java (hunk 7 above)
Comment (Rachael-E) [reply utility_network/perform-valve-isolation-trace/src/main/java/com/esri/samples/perform_valve_isolation_trace/PerformValveIsolationTraceController.java:386]: from the sample design, yeh that's the one we want! thanks for checking :)

**Pull request merged**